platform/riscv-virt/pci.c

```diff
@@ -0,0 +1,151 @@
+#include <kernel.h>
+#include <pci.h>
+#include <drivers/console.h>
+
+//#define PCI_PLATFORM_DEBUG
+#ifdef PCI_PLATFORM_DEBUG
+#define pci_plat_debug rprintf
+#else
+#define pci_plat_debug(...) do { } while(0)
+#endif
+
+#define PIO_DATA (mmio_base_addr(PCIE_PIO))
+#define pio_in8(port) mmio_read_8(PIO_DATA + port)
+#define pio_in16(port) mmio_read_16(PIO_DATA + port)
+#define pio_in32(port) mmio_read_32(PIO_DATA + port)
+#define pio_out8(port, source) mmio_write_8(PIO_DATA + port, source)
+#define pio_out16(port, source) mmio_write_16(PIO_DATA + port, source)
+#define pio_out32(port, source) mmio_write_32(PIO_DATA + port, source)
+
+/* stub ... really shouldn't be hardwired into console.c */
+void vga_pci_register(kernel_heaps kh, console_attach a)
+{
+}
+
+u32 pci_cfgread(pci_dev dev, int reg, int bytes)
```

Comment:
If you don't mind, I will defer this for now and handle it in a separate change after Francesco gets the Graviton changes in, since that changes the way some of this works.